signer/src/testing/context.rs

```diff
@@ -184,6 +248,327 @@ impl BitcoinInteract for WrappedMock<MockBitcoinInteract> {
+        self.inner.lock().await.get_block(block_id).await
+    }
+
+    async fn get_tenure(&self, block_id: StacksBlockId) -> Result<Vec<NakamotoBlock>, Error> {
+        self.inner.lock().await.get_tenure(block_id).await
+    }
+
+    async fn get_tenure_info(&self) -> Result<RPCGetTenureInfo, Error> {
+        self.inner.lock().await.get_tenure_info().await
+    }
+
+    async fn estimate_fees<T>(&self, payload: &T, priority: FeePriority) -> Result<u64, Error>
+    where
+        T: AsTxPayload + Send + Sync,
+    {
+        self.inner
+            .lock()
+            .await
+            .estimate_fees(payload, priority)
+            .await
+    }
+
+    fn nakamoto_start_height(&self) -> u64 {
+        tokio::runtime::Handle::current().block_on(async move {
+            let inner = self.inner.lock().await;
+            inner.nakamoto_start_height()
+        })
+    }
+}
+
+/// Struct which holds the current configuration of the context builder.
+pub struct ContextConfig<Storage, Bitcoin, Stacks> {
+    settings: crate::config::Settings,
+    storage: Storage,
+    bitcoin: Bitcoin,
+    stacks: Stacks,
+}
+
+impl Default for ContextConfig<(), (), ()> {
+    fn default() -> Self {
+        Self {
+            settings: Settings::new_from_default_config().expect("failed to load default config"),
+            storage: (),
+            bitcoin: (),
+            stacks: (),
+        }
+    }
+}
+
+/// State for the builder pattern.
+pub trait BuilderState<Storage, Bitcoin, Stacks> {
+    /// Consumes the builder, returning its current internal configuration.
+    fn get_config(self) -> ContextConfig<Storage, Bitcoin, Stacks>;
+}
+
+/// A builder for creating a [`TestContext`].
+pub struct ContextBuilder<Storage, Bitcoin, Stacks> {
+    config: ContextConfig<Storage, Bitcoin, Stacks>,
+}
+
+impl ContextBuilder<(), (), ()> {
+    /// Create a new context builder.
+    pub fn new() -> Self {
+        Self { config: Default::default() }
+    }
+}
+
+impl Default for ContextBuilder<(), (), ()> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<Storage, Bitcoin, Stacks> BuilderState<Storage, Bitcoin, Stacks>
+    for ContextBuilder<Storage, Bitcoin, Stacks>
+{
+    fn get_config(self) -> ContextConfig<Storage, Bitcoin, Stacks> {
+        self.config
+    }
+}
+
+/// Trait for configuring the settings. These methods are always available.
+pub trait ConfigureSettings<Storage, Bitcoin, Stacks>
+where
+    Self: Sized + BuilderState<Storage, Bitcoin, Stacks>,
+{
+    /// Configure the context with the specified settings.
+    fn with_settings(self, settings: Settings) -> ContextBuilder<Storage, Bitcoin, Stacks> {
+        let config = self.get_config();
+        ContextBuilder {
+            config: ContextConfig {
+                settings,
+                storage: config.storage,
+                bitcoin: config.bitcoin,
+                stacks: config.stacks,
+            },
```

Comment:
nit: here and below, use struct update syntax?

Reply:
can't due to generics

Reply:
or actually this specific one can since we're not changing the type, but the other's cant. updated this one in [f4951ef](https://github.com/stacks-network/sbtc/pull/606/commits/f4951ef29f8ea66b9a3c2ea436ff0d668395b9fa)